src/helper/sampler.hpp

```diff
@@ -0,0 +1,102 @@
+// Unless explicitly stated otherwise all files in this repository are
+// dual-licensed under the Apache-2.0 License or BSD-3-Clause License.
+//
+// This product includes software developed at Datadog
+// (https://www.datadoghq.com/). Copyright 2022 Datadog, Inc.
+
+#pragma once
+
+#include <atomic>
+#include <cmath>
+#include <cstdint>
+#include <iostream>
+#include <mutex>
+#include <optional>
+
+namespace dds {
+static const double min_rate = 0.0001;
+class sampler {
+public:
+    sampler(double sample_rate)
+    {
+        // NOLINTBEGIN(cppcoreguidelines-avoid-magic-numbers,readability-magic-numbers)
+        if (sample_rate <= 0) {
+            sample_rate = 0;
+        } else if (sample_rate > 1) {
+            sample_rate = 1;
+        } else if (sample_rate < min_rate) {
+            sample_rate = min_rate;
+        }
+
+        group_size_ = static_cast<unsigned>(round(1 / sample_rate));
```

Comment:
what if it's larger than .5? Plus, this seems very rough for anything but small values of `sample_rate`.

Reply:
I have changed the implementation so this is not relevant anymore